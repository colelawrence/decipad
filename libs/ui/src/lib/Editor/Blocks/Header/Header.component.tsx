import { Text } from '@chakra-ui/react';
import React from 'react';
import { RenderElementProps } from 'slate-react';

interface HeaderProps extends RenderElementProps {
  size?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
}

export const Header = ({
  attributes,
  children,
  size = 'h1',
}: HeaderProps): JSX.Element => {
  switch (size) {
    case 'h2':
      return (
        <Text as="h2" fontSize="3xl" fontWeight="bold" {...attributes}>
          {children}
        </Text>
      );
    case 'h3':
      return (
        <Text as="h2" fontSize="2xl" fontWeight="bold" {...attributes}>
          {children}
        </Text>
      );
    case 'h4':
      return (
        <Text as="h2" fontSize="xl" fontWeight="bold" {...attributes}>
          {children}
        </Text>
      );
    case 'h5':
      return (
        <Text as="h2" fontSize="lg" fontWeight="bold" {...attributes}>
          {children}
        </Text>
      );
    case 'h6':
      return (
        <Text as="h2" fontSize="md" fontWeight="bold" {...attributes}>
          {children}
        </Text>
      );
    default:
      return (
        <Text as="h1" fontSize="4xl" fontWeight="bold" {...attributes}>
          {children}
        </Text>
      );
  }
};
