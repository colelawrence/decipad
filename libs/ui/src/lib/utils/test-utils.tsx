import { ChakraProvider } from '@chakra-ui/react';
import { render, RenderOptions } from '@testing-library/react';
import React from 'react';
import { theme } from '../theme';

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <ChakraProvider resetCSS theme={theme}>
      {children}
    </ChakraProvider>
  );
};

const customRender = (
  ui: JSX.Element,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
