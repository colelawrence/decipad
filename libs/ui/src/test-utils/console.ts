export const mockConsoleWarn = () => {
  const originalConsoleWarn = console.warn;
  const mockConsoleWarn: jest.MockedFunction<typeof console.warn> = jest.fn();

  beforeEach(() => {
    console.warn = jest.fn();
    mockConsoleWarn.mockClear();
  });
  afterEach(() => {
    console.warn = originalConsoleWarn;
  });

  return mockConsoleWarn;
};
