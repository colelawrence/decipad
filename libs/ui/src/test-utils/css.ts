import cssVarsPonyfill from 'css-vars-ponyfill';

export const findParentWithStyle = <P extends keyof CSSStyleDeclaration>(
  element: Element | null,
  propertyName: P
):
  | (Pick<CSSStyleDeclaration, P> & {
      element: Element;
      styles: CSSStyleDeclaration;
    })
  | null => {
  if (element === null) {
    return null;
  }
  const styles = getComputedStyle(element);
  if (!styles[propertyName]) {
    return findParentWithStyle(element.parentElement, propertyName);
  }
  return {
    [propertyName]: styles[propertyName],
    element,
    styles,
    // cannot type dynamic property key based on type parameter
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any;
};

export const applyCssVars = (): Promise<() => void> => {
  return new Promise((resolve, reject) => {
    const cleanup = () =>
      [...document.head.querySelectorAll('[data-cssvars="out"]')].forEach(
        (element) => element.remove()
      );
    cssVarsPonyfill({
      onError: reject,
      onComplete: () => resolve(cleanup),
    });
  });
};

export const applyPrefersColorScheme = async (
  colorScheme: 'dark' | 'light'
): Promise<() => void> => {
  // JSDOM ignores CSSOM modifications, so build a CSS string and create and append a style element instead
  const styleText = [...document.styleSheets]
    .flatMap((styleSheet) => [...styleSheet.cssRules])
    .flatMap((rule) => {
      if (rule.type === CSSRule.MEDIA_RULE) {
        const mediaRule = rule as CSSMediaRule;
        for (
          let mediaIndex = 0;
          mediaIndex < mediaRule.media.length;
          mediaIndex += 1
        ) {
          if (
            mediaRule.media[mediaIndex].replace(/\s/g, '') ===
            `(prefers-color-scheme:${colorScheme})`
          ) {
            return [...mediaRule.cssRules];
          }
        }
      }
      return [];
    })
    .flatMap((rule) => rule.cssText)
    .join(' ');

  const styleElement = document.createElement('style');
  styleElement.innerHTML = styleText;
  document.head.appendChild(styleElement);

  const cleanupCssVars = await applyCssVars();

  return () => {
    cleanupCssVars();
    styleElement.remove();
  };
};
