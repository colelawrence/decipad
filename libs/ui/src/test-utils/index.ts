export * from './menu';
export * from './language';
