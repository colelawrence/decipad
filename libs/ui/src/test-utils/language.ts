import {
  Result,
  runCode as run,
  SerializedTypeKind,
  serializeResult,
} from '@decipad/computer';

export async function runCode<T extends SerializedTypeKind>(
  code: string
): Promise<Result.Result<T>> {
  const result = await run(code);
  return serializeResult<T>(result.type, result.value);
}
