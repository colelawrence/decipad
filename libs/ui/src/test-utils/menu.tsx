import { FC } from 'react';
import * as RadixDropdown from '@radix-ui/react-dropdown-menu';

export const MenuWrapper: FC<React.PropsWithChildren<unknown>> = ({
  children,
}) => (
  <RadixDropdown.Root open>
    <RadixDropdown.Trigger css={{ height: 0 }}>&nbsp;</RadixDropdown.Trigger>
    <RadixDropdown.Content>{children}</RadixDropdown.Content>
  </RadixDropdown.Root>
);
