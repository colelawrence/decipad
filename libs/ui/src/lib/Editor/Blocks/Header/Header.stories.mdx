import { Meta, Story, Canvas, Props } from '@storybook/addon-docs/blocks';
import { Header } from './Header.component';

<Meta
  title="Legacy/Editor/Blocks/Header"
  parameters={{
    docs: {
      source: {
        code: `<Header
  element={{
    children: [],
  }}
  attributes={{ 'data-slate-node': 'element', ref: null }}
>
  Hello World
</Header>`,
      },
    },
  }}
  argTypes={{
    children: {
      defaultValue: 'Hello World',
      description: 'The text content of the component',
      type: 'string',
    },
    size: {
      defaultValue: 'h1',
      description: 'The size of the header',
      options: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'],
      type: 'select',
    },
  }}
  component={Header}
/>

# Header

The Header block component is the slate component that renders when the user wants to input a header, there are six options for the size of the header.

# Available Sizes

- h1
- h2
- h3
- h4
- h5
- h6

export const Template = (args) => <Header {...args} />;

<Canvas>
  <Story name="Header">{Template.bind({})}</Story>
</Canvas>

<Props story="Header" />
