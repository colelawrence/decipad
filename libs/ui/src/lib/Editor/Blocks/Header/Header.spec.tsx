import React from 'react';
import { render } from 'test-utils';
import { Header } from './Header.component';

describe('Header Block', () => {
  it('renders', () => {
    const { container } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        Hello World
      </Header>
    );

    expect(container).toBeDefined();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        Hello World
      </Header>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders the correct h1 size', () => {
    const { getByText } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        Hello World
      </Header>
    );

    expect(getByText('Hello World')).toHaveStyleRule(
      'font-size',
      'var(--chakra-fontSizes-4xl)'
    );
  });

  it('renders the correct h2 size', () => {
    const { getByText } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
        size="h2"
      >
        Hello World
      </Header>
    );

    expect(getByText('Hello World')).toHaveStyleRule(
      'font-size',
      'var(--chakra-fontSizes-3xl)'
    );
  });

  it('renders the correct h3 size', () => {
    const { getByText } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
        size="h3"
      >
        Hello World
      </Header>
    );

    expect(getByText('Hello World')).toHaveStyleRule(
      'font-size',
      'var(--chakra-fontSizes-2xl)'
    );
  });

  it('renders the correct h4 size', () => {
    const { getByText } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
        size="h4"
      >
        Hello World
      </Header>
    );

    expect(getByText('Hello World')).toHaveStyleRule(
      'font-size',
      'var(--chakra-fontSizes-xl)'
    );
  });

  it('renders the correct h5 size', () => {
    const { getByText } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
        size="h5"
      >
        Hello World
      </Header>
    );

    expect(getByText('Hello World')).toHaveStyleRule(
      'font-size',
      'var(--chakra-fontSizes-lg)'
    );
  });

  it('renders the correct h6 size', () => {
    const { getByText } = render(
      <Header
        element={{
          children: [{ text: 'Hello World' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
        size="h6"
      >
        Hello World
      </Header>
    );

    expect(getByText('Hello World')).toHaveStyleRule(
      'font-size',
      'var(--chakra-fontSizes-md)'
    );
  });
});
