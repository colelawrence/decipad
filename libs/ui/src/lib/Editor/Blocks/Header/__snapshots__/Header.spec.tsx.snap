// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Block matches snapshot 1`] = `
.emotion-0 {
  font-size: var(--chakra-fontSizes-4xl);
  font-weight: var(--chakra-fontWeights-bold);
}

<div>
  <h1
    class="chakra-text emotion-0"
    data-slate-node="element"
  >
    Hello World
  </h1>
</div>
`;
